hardware/ctre/src/main/java/io/github/frc5024/lib5k/hardware/ctre/sensors/ExtendedPigeonIMU.java

```diff
@@ -0,0 +1,124 @@
+
+import io.github.frc5024.lib5k.hardware.common.drivebase.IDifferentialDrivebase;
+import io.github.frc5024.lib5k.hardware.common.sensors.interfaces.ISimGyro;
+import io.github.frc5024.lib5k.hardware.common.sensors.simulation.GyroSimUtil;
+
+/**
+ * A wrapper for the CTRE PigeonIMU that brings it into the 5024 ecosystem.
+ * 
+ * http://www.ctr-electronics.com/gadgeteer-imu-module-pigeon.html#product_tabs_technical_resources
+ */
+public class ExtendedPigeonIMU extends PigeonIMU implements ISimGyro {
+
+    // Inversion tracker
+    private boolean inverted = false;
+    private boolean calibrated = false;
+
+    // Simulation
+    private GyroSimUtil sim;
+
+    public ExtendedPigeonIMU(int canID) {
+        super(canID);
+        super.configFactoryDefault();
+    }
+
+    public ExtendedPigeonIMU(TalonSRX talonSrx) {
+        super(talonSrx);
+        super.configFactoryDefault();
+    }
+
+    @Override
+    public void setInverted(boolean inverted) {
+        if (sim != null && sim.simReady()) {
+            sim.setInverted(inverted);
+        }
+        this.inverted = inverted;
+    }
+
+    @Override
+    public boolean getInverted() {
+        if (sim != null && sim.simReady()) {
+            return sim.getInverted();
+        }
+        return inverted;
+    }
+
+    @Override
+    public double getHeading() {
+        return Math.IEEEremainder(getAngle(), 360) * (inverted ? -1.0 : 1.0);
+    }
+
+    @Override
+    public void calibrate() {
+        if (sim != null && sim.simReady()) {
+            sim.calibrate();
+        } else {
+            //* Note: The PigeonIMU is not designed to be calibrated on-robot
+            //* super.enterCalibrationMode(CalibrationMode.BootTareGyroAccel);
+        }
+        calibrated = true;
+
+    }
+
+    @Override
+    public boolean getCalibrated() {
+        return calibrated;
+    }
+
+    @Override
+    public void reset() {
+        if (sim != null && sim.simReady()) {
+            sim.reset();
+        } else {
+            // CTRE doesn't really provide a reset for us to use
+            super.setCompassAngle(0.0);
+        }
+
+    }
+
+    @Override
+    public double getAngle() {
+        if (sim != null && sim.simReady()) {
+            return sim.getAngle();
+        }
+        return super.getCompassHeading() * ((inverted) ? -1 : 1);
+    }
+
+    @Override
+    public double getRate() {
+        if (sim != null && sim.simReady()) {
+            return sim.getRate();
+        }
+
+        // Build a data frame array
+        double[] data = new double[3];
+
+        // Read a data frame from sensor
+        super.getRawGyro(data);
+
+        // Return the Z axis in dps
+        return data[2] * ((inverted) ? -1 : 1);
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (sim != null) {
+            sim.close();
+        }
+
+    }
+
+    @Override
+    public void initDrivebaseSimulation(IDifferentialDrivebase drivebase) {
```

Comment:
Maybe make a constructor that has an option to call this immediately.

Reply:
It is designed to be called this way, from inside DriveTrain:
```java
public DriveTrain(){
	if(RobotBase.isSimulation()){
		gyro.initDrivebaseSimulation(this);
	}
}
```